src/enzo/ActiveParticle.h

```diff
@@ -0,0 +1,816 @@
+  /* For every entry in the ActiveParticleFormationData struct, we
+   * have a bool here. */
+  bool DarkMatterDensity;
+  bool H2Fraction;
+  bool CoolingTime;
+  bool CoolingRate;
+  bool Temperature;
+  bool MetalField;
+};
+
+const struct ActiveParticleFormationDataFlags flags_default = {
+  false,    // DarkMatterDensity
+  false,    // H2Fraction
+  false,    // CoolingTime
+  false,    // CoolingRate
+  false,    // Temperature
+  false     // MetalField
+};
+
+namespace ActiveParticleHelpers {
+
+  template <class APClass> int CalculateElementSize() {
+      static int particle_size = 0;
+      if (particle_size > 0) return particle_size;
+      AttributeVector &handlers = APClass::AttributeHandlers;
+      for(AttributeVector::iterator it = handlers.begin();
+          it != handlers.end(); ++it) {
+          particle_size += (**it).element_size;
+      }
+      return particle_size;
+  }
+
+  template <class APClass> void Allocate(int Count, char **buffer) {
+
+      /* This routine is called for each particle type. */
+      /* So we need to re-calculate the element and header size for each. */
+
+      int particle_size = CalculateElementSize<APClass>();
+      int header_size = sizeof(int);
+
+      *buffer = new char[particle_size * Count + header_size];
+
+  }
+
+  template <class APClass> void PrintActiveParticle(APClass *ap,
+      const std::string prefix = "") {
+
+      AttributeVector &handlers = APClass::AttributeHandlers;
+      for(AttributeVector::iterator it = handlers.begin();
+        it != handlers.end(); ++it) {
+          (*it)->PrintAttribute(ap);
+          std::cout << " ";
+      }
+
+      std::cout << std::endl;
+
+  }
+
+  template <class APClass> void WriteParticles(
+          ActiveParticleList<ActiveParticleType>& InList, int ParticleTypeID,
+          int TotalParticles,
+          const std::string name, hid_t grid_node)
+  {
+      int i, size = 0, Count = 0;
+      char *buffer, *_buffer;
+      AttributeVector &handlers = APClass::AttributeHandlers;
+      const char *_ap_name = name.c_str();
+      hid_t node = H5Gcreate(grid_node, _ap_name, 0);
+
+      /* Now we count up our particles */
+      for (i = 0; i < TotalParticles; i++) {
+        if (InList[i]->GetEnabledParticleID() == ParticleTypeID) Count++;
+      }
+
+      writeScalarAttribute(node, HDF5_INT, "Count", &Count);
+
+      if(Count == 0) {
+        H5Gclose(node);
+        return;
+      }
+
+      int ndims = 1;
+      hsize_t dims[2] = {1, 1};
+      //const int NormCount = Count;
+      APClass *In;
+
+      for (AttributeVector::iterator it = handlers.begin();
+          it != handlers.end(); ++it) {
+	  const char *_name = (*it)->name.c_str();
+	  /* HACK - there must be a better way to do this...*/
+	  if( (strcmp(_name, "AccretionRate") == 0 ||
+	       strcmp(_name, "AccretionRateTime") == 0 ||
+	       strcmp(_name, "Accreted_angmom") == 0)
+	      && strcmp(_ap_name, "SmartStar") == 0)
+	    {
```

Comment:
This looks very problem specific hack and a more generalized way of doing this must be implemented before merging.

Reply:
Agreed! Lets come back to this.

Reply:
Since it works for now, we won't fix this for the PR. But it should be fixed in the future, so I created issue #74.